text/3723-static-function-argument-unpacking.md

```diff
@@ -0,0 +1,719 @@
+- Tuples: [stdlib tuple primitive](https://doc.rust-lang.org/std/primitive.tuple.html).
+- Arrays: [stdlib array primitive](https://doc.rust-lang.org/std/primitive.array.html).
+
+Various Rust books may want to mention or teach the feature. For example, The Rust Programming Language book's [Appendix B: Operators and Symbols](https://doc.rust-lang.org/book/appendix-02-operators.html) could include the syntax.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Functions that accept many parameters may already be a code smell, and the proposed change would likely help calling such functions the most, becoming an enabler for anti-patterns. At the same time, unpacking three of four arguments by hand is not much work, which means that the impact of the change in normal code is not huge.
+
+A sufficiently smart language server could automate argument unpacking, slightly decreasing the usefulness of having the feature in language itself when writing new code. However, there are many scenarios where a language server doesn't help, such as code examples in books.
+
+Although the proposed syntax is familiar from other contexts, e.g. from other programming languages, it still burdens developers with additional syntax to understand. Possibly, depending on how intuitive the syntax is or how familiar the developer is with similar features from other programming languages, this may or may not imply an additional mental overhead when working with Rust code. However, as the new syntax comes in the form of syntactic sugar, this shouldn't be so bad: no-one is forced to use this even though they may be forced to understand this when reading code written by others. Additionally, it could be reasonably argued that the proposed change makes the language a bit more consistent, since a similar feature for struct instantiation already exists. Anecdotally, the author of this RFC tried to use the syntax for the proposed feature only to notice it doesn't exist yet.
+
+Any initiatives for the distinct features of *named parameters*, *optional/variadic parameters*, *parameters with default values* or combinations thereof will need to consider the corresponding proposals' interactions with *argument unpacking*. The selected syntax of `...` will also be cemented to specific uses, possibly denying its use in some other contexts.
+
+Ecosystem churn with MSRV *(Minimal Supported Rust Version)* bumps may be expected as some crate authors may decide to use argument unpacking in places where a workaround was previously used.
+
+The ellipsis symbol composed from three consecutive ASCII dot characters is used in the "et cetera" or "and so on" sense in many design documents and code examples. Giving it an actual syntactical meaning could lead to some confusion or readability issues. Preferring `…`, i.e. the Unicode character U+2026, Horizontal Ellipsis, in those places could help.
+
+# Rationale and Alternatives
```

Comment:
There's a bit about this under "Future Possibilities" subchapter "Unpacking in Fixed-Size Array and Tuple Literals", but I should make it more general and expand it to cover additional cases as well – such as that char array example you gave.

Reply:
Ultimately, I'd love to see unpacking expanded thus.

The reason I left it out of scope for now is that, presumably, non-trivial additional work would be required to think over the possible interactions with other features and explore the design space thoroughly enough.

My *preference* would be to postpone expansion of the feature this way into a future RFC, but I'm not strongly opposed to including it in this one. I think postponing would also have the benefit, that – assuming this current RFC gets accepted – the RFC would be more lightweight. Either way, maybe I should elaborate on the reason for omission from the RFC.

What's your view on this – include now or have a separate RFC?

Reply:
I added further elaboration on the subject of unpacking within collection literals in commit 50339174ee755122bc5076c23ef4aad1ebc88a06. There's an IRLO thread from the end of 2020 with a [pre-RFC on array expansion syntax](https://internals.rust-lang.org/t/pre-rfc-array-expansion-syntax/13490) by nwn that pretty much covers this idea, but would need synchronization on the selected syntax. (I wonder if they'd like to continue on this by themselves or work together to write a new version?)

Reply:
I've also written a bit more about the limited scope into the RFC in commit 3a2f4500792a56c364551b61d8499eac916372a0. FWIW, I'd be happy to write full RFCs for some of the listed future possibilities expanding on this proposal, provided that this one has a good chance of ultimately getting accepted. :)


<details>
    <summary>Some general thoughts on "minimum viable RFCs", somewhat offtopic maybe, take as you will</summary>

---

As I see it, planning, writing, sharing, and discussing major initiatives such as this RFC requires a non-trivial investment of time – don't get me wrong, so far I'm enjoying this. This is why I think suggesting a bite-sized proposal at a time to gauge interest works well from an RFC author's perspective and decreases any risks on burnout or sadness stemming from seeing the work not receiving interest or getting rejected (or at worst, getting ridiculed, but I'm sure the risks for that in the Rust community are very low to begin with).
</details>

Reply:
One possibility for a way forward that just occurred to me is landing these connected features (with accepted RFCs) into nightly one by one, but stabilizing them in one batch. This way, users of stable Rust would benefit from the features without any of them feeling unfinished.

Reply:
I think not including tuple, array, and slice constructors in the MVP is a mistake. Users will be very surprised to learn that this syntax works with tuple structs and not normal tuples. Also, making this syntax exclusive to tuple structs will incentivize developers to use them even when a tuple or array would otherwise be better suited (including in public APIs). (I agree with all the other future-possibility relegations.)

Reply:
I agree on tuples and arrays. But I think that slices has enough additional complexity it is worth excluding from the  MVP.

Reply:
To be clear, I was only referring to unpacking *into* slices, not unpacking slices. But now that I think about it more, what that really means is unpacking into arrays that are immediately reborrowed as slices. So yes, tuples and arrays.